PhysicsTools/PatAlgos/plugins/PATLostTracks.cc

```diff
@@ -255,16 +271,84 @@ void pat::PATLostTracks::addPackedCandidate(std::vector<pat::PackedCandidate>& c
+
+  if (trk->pt() > minPtToStoreProps_ || trkStatus == TrkStatus::VTX) {
+    if (std::abs(id) == 11 || trkStatus == TrkStatus::VTX) {
+      cands.back().setTrackProperties(*trk, covariancePackingSchemas_[4], covarianceVersion_);
+    } else {
+      if (trk->hitPattern().numberOfValidPixelHits() > 0) {
+        cands.back().setTrackProperties(
+            *trk, covariancePackingSchemas_[0], covarianceVersion_);  // high quality with pixels
+      } else {
+        cands.back().setTrackProperties(
+            *trk, covariancePackingSchemas_[1], covarianceVersion_);  // high quality without pixels
+      }
+    }
+  } else if (trk->pt() > 0.5) {
+    if (trk->hitPattern().numberOfValidPixelHits() > 0) {
+      cands.back().setTrackProperties(
+          *trk, covariancePackingSchemas_[2], covarianceVersion_);  // low quality with pixels
+    } else {
+      cands.back().setTrackProperties(
+          *trk, covariancePackingSchemas_[3], covarianceVersion_);  // low quality without pixels
+    }
+  }
+  cands.back().setAssociationQuality(pvAssocQuality);
+}
+
+std::pair<int, pat::PackedCandidate::PVAssociationQuality> pat::PATLostTracks::associateTrkToVtx(
+    const reco::VertexCollection& vertices, const reco::TrackRef& trk) const {
+  //Inspired by CommonTools/RecoAlgos/interface/PrimaryVertexAssignment.h
+  //but without specific association for secondaries in jets and option to use timing
 
-  if (trk->pt() > minPtToStoreProps_ || trkStatus == TrkStatus::VTX)
-    cands.back().setTrackProperties(*trk, covarianceSchema_, covarianceVersion_);
-  if (pvOrig.trackWeight(trk) > 0.5) {
-    cands.back().setAssociationQuality(pat::PackedCandidate::UsedInFitTight);
+  int iVtxMaxWeight = -1;
+  int iVtxMinDzDist = -1;
+  size_t idx = 0;
+  float maxWeight = 0;
+  double minDz = std::numeric_limits<double>::max();
+  double minDzSig = std::numeric_limits<double>::max();
+  for (auto const& vtx : vertices) {
+    float w = vtx.trackWeight(trk);
+    double dz = std::abs(trk->dz(vtx.position()));
+    double dzSig = dz / trk->dzError();
+    if (w > maxWeight) {
+      maxWeight = w;
+      iVtxMaxWeight = idx;
+    }
+    if (dzSig < minDzSig) {
+      minDzSig = dzSig;
+      minDz = dz;
+      iVtxMinDzDist = idx;
+    }
+    idx++;
+  }
+  // vertex in which fit the track was used
+  if (iVtxMaxWeight >= 0) {
+    if (maxWeight > 0.5) {
+      return std::pair<int, pat::PackedCandidate::PVAssociationQuality>(iVtxMaxWeight,
+                                                                        pat::PackedCandidate::UsedInFitTight);
+    } else {
+      return std::pair<int, pat::PackedCandidate::PVAssociationQuality>(iVtxMaxWeight,
+                                                                        pat::PackedCandidate::UsedInFitLoose);
+    }
+  }
+  // vertex "closest in Z" with tight cuts (targetting primary particles)
+  const double add_cov = iVtxMinDzDist >= 0 ? vertices[iVtxMinDzDist].covariance(2, 2) : 0.f;
```

Comment:
The computation of `add_cov` and `dzErr` can be restricted to the cases in which `iVtxMinDzDist >= 0 && minDz < maxDzForPrimaryAssignment_`, i.e. the first condition of the following `if` can be better factorized out

Reply:
(Well, `minDz < maxDzForPrimaryAssignment_` can be enough to ensure that also` iVtxMinDzDist >= 0`)

Reply:
It will be done